File: 019.RemoveNthNode/wyrj/RemoveNthNode.c
/* 0ms */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode *start = head, *end = head;
    while(n--) {
        end = end->next;
    }
    if (NULL == end) {
        return (head) ? head->next : NULL;
    }
    while(NULL != end->next) {
        start = start->next;
        end = end->next;
    }
    start->next = start->next->next;
    return head;
}

File: 019.RemoveNthNode/Ddfish/RemoveNthNode.swift

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

class Solution {
    func removeNthFromEnd(_ head: ListNode?, _ n: Int) -> ListNode? {
        
        let tempNode = ListNode(0)
        tempNode.next =  head
        
        var first  : ListNode? = tempNode
        var second : ListNode? = tempNode
        
        var count = 0
        while first != nil {
            first = first?.next
            if(count > n){
                second = second?.next
            }
            count += 1
        }
        second?.next = second?.next?.next
        
        
        return tempNode.next
    }
}
